frb_example/pure_dart/dart/lib/main.dart

```diff
@@ -256,28 +256,28 @@ void main(List<String> args) async {
   });
 
   test('dart call handleEnumStruct', () async {
-    expect(await api.handleEnumStruct(val: Empty()), Empty());
+    expect(await api.handleEnumStruct(val: KitchenSink_Empty_Variant()), KitchenSink_Empty_Variant());
```

Comment:
As discussed before, shall we use `KitchinSink_Empty` instead of `KitchenSink_Empty_Variant`

Reply:
I agree we can come up with a better naming around structs generated for enums variants.

The issue with name like e.g. `KitchenSink_Empty` is that it already conflicts with pre-existing generated code.

```bash
➜  flutter_rust_bridge git:(fix/enum-variant-naming-conflicts) ✗ just test-pure                
cd frb_example/pure_dart/rust && cargo b
   Compiling flutter_rust_bridge_example_single_block_test v1.0.0 (/Users/romain/Development/contributions/flutter_rust_bridge/frb_example/pure_dart/rust)
    Finished dev [unoptimized + debuginfo] target(s) in 0.82s
cd frb_example/pure_dart/dart && dart pub get && dart lib/main.dart ../rust/target/debug/libflutter_rust_bridge_example.dylib
Resolving dependencies... (1.2s)
Got dependencies!
lib/bridge_generated.freezed.dart:200:16: Error: 'API_Speed' is already declared in this scope.
abstract class API_Speed implements API {
               ^^^^^^^^^
lib/bridge_generated.dart:4292:7: Context: Previous declaration of 'API_Speed'.
class API_Speed extends ffi.Struct {
      ^^^^^^^^^
lib/bridge_generated.freezed.dart:338:16: Error: 'API_KilometerPoint' is already declared in this scope.
abstract class API_KilometerPoint implements API {
               ^^^^^^^^^^^^^^^^^^
lib/bridge_generated.dart:4316:7: Context: Previous declaration of 'API_KilometerPoint'.
class API_KilometerPoint extends ffi.Struct {
      ^^^^^^^^^^^^^^^^^^
lib/bridge_generated.freezed.dart:960:16: Error: 'KilometerPoint_Unknown' is already declared in this scope.
abstract class KilometerPoint_Unknown implements KilometerPoint {
               ^^^^^^^^^^^^^^^^^^^^^^
lib/bridge_generated.dart:4296:7: Context: Previous declaration of 'KilometerPoint_Unknown'.
class KilometerPoint_Unknown extends ffi.Opaque {}
      ^^^^^^^^^^^^^^^^^^^^^^
lib/bridge_generated.freezed.dart:1085:16: Error: 'KilometerPoint_SNCF' is already declared in this scope.
abstract class KilometerPoint_SNCF implements KilometerPoint {
               ^^^^^^^^^^^^^^^^^^^
lib/bridge_generated.dart:4298:7: Context: Previous declaration of 'KilometerPoint_SNCF'.
class KilometerPoint_SNCF extends ffi.Struct {
      ^^^^^^^^^^^^^^^^^^^
lib/bridge_generated.freezed.dart:1294:16: Error: 'KitchenSink_Empty' is already declared in this scope.
abstract class KitchenSink_Empty implements KitchenSink {
               ^^^^^^^^^^^^^^^^^
lib/bridge_generated.dart:4177:7: Context: Previous declaration of 'KitchenSink_Empty'.
class KitchenSink_Empty extends ffi.Opaque {}
      ^^^^^^^^^^^^^^^^^
error: Recipe `test-pure` failed on line 46 with exit code 254
```

Any other convention in mind ? @fzyzcjy

Reply:
Ah that sounds like another problem. The https://github.com/fzyzcjy/flutter_rust_bridge/blob/386a20f1c70a4ac76ce78b129f0f8a35ec39daff/frb_example/pure_dart/dart/lib/bridge_generated.dart#L4292 "API_Speed" should have been "wire_API_speed" imho, since by convention, all those are wire structs and should prefix like that.

/cc @Desdaemon  who implemented the great enhanced enum feature

Reply:
Oh my latest commit renamed the enums ... :man_facepalming: sorry it's gonna make review more complicated here.

Reply:
I don't mind doing any change, just waiting for your recommendations @fzyzcjy @Desdaemon

Reply:
Personally I suggest the wire struct to have name `wire_...` to be consistent. Waiting for @Desdaemon 's opinion

Reply:
Given @Desdaemon has not replied yet, I guess he may not have strong disagreement about this topic :)

Reply:
You know what, I'll try to get the refactoring into `wire_...` done by Monday on a separate branch so you guys both have a chance to see if it's what you want or not :)

Reply:
Take your time :) IMHO it is also completely ok to directly write it in this branch since that simplifies coding

Reply:
I must admit I kinda lost the thread of which method generates what inside codegen, so I'm having a hard time finding where the changes should be made @fzyzcjy

Reply:
furthermore, I think it's partly already the case, looking at the current state of the code, `wire_${enum_name}` is already generated :

```dart
// bridge_generated.dart
void _api_fill_to_wire_measure(Measure apiObj, wire_Measure wireObj) {
    if (apiObj is Measure_Speed_Variant) {
      wireObj.tag = 0;
      wireObj.kind = inner.inflate_Measure_Speed();
      wireObj.kind.ref.Speed.ref.field0 = _api2wire_box_speed(apiObj.field0);
    }
    if (apiObj is Measure_Distance_Variant) {
      wireObj.tag = 1;
      wireObj.kind = inner.inflate_Measure_Distance();
      wireObj.kind.ref.Distance.ref.field0 = _api2wire_box_distance(apiObj.field0);
    }
  }
```
where `wire_Measure` is:
```dart
// bridge_generated.dart
class wire_Measure extends ffi.Struct {
  @ffi.Int32()
  external int tag;

  external ffi.Pointer<MeasureKind> kind;
}
```
and `Measure` is:
```dart
// bridge_generated.dart
@freezed
class Measure with _$Measure {
  const factory Measure.speed(
    Speed field0,
  ) = Measure_Speed_Variant;
  const factory Measure.distance(
    Distance field0,
  ) = Measure_Distance_Variant;
}
```
So I'm reluctant to process this change as I think it's gonna require more cascading changes and increase the likeness of me introducing bugs/regressions.

Would you mind me leaving it the way it currently is ? @fzyzcjy

Reply:
> lost the thread of which method generates what inside codegen, so I'm having a hard time finding where the changes should be made

Oops... What should I do? Shall I add a bit of doc or comments?

Btw I guess most generation are self-explanatory by looking at names.

This new chapter may be helpful: http://cjycode.com/flutter_rust_bridge/contributing/design.html

Reply:
I see. So `Measure` seems like the struct that users are going to use, and `wire_Measure` is the one that we use internally on the "wire".

So, in the Measure example above, maybe we should simply rename Measure_Speed_Variant to Measure_Speed etc.  i.e. just simply remove "_Variant"

If this is hard, leaving the "_Variant" there looks also acceptable.

Reply:
> So, in the Measure example above, maybe we should simply rename Measure_Speed_Variant to Measure_Speed etc. i.e. just simply remove "_Variant"
> 
> If this is hard, leaving the "_Variant" there looks also acceptable.

Yeah that bring us back to [this previous comment](https://github.com/fzyzcjy/flutter_rust_bridge/pull/604#discussion_r935045259) in the discussion.
If we strip `_Variant` suffix, then it also conflicts with pre-existing names.

Reply:
Do we still have this problem? Looking at 

https://github.com/fzyzcjy/flutter_rust_bridge/blob/e63577dd126c03405bd207ec153aad222702cbce/frb_example/pure_dart/dart/lib/bridge_generated.dart
https://github.com/fzyzcjy/flutter_rust_bridge/blob/e63577dd126c03405bd207ec153aad222702cbce/frb_example/pure_dart/dart/lib/bridge_generated.freezed.dart

For "Measure" I see:

```
@freezed
class Measure with _$Measure {
  const factory Measure.speed(
    Speed field0,
  ) = Measure_Speed_Variant;
  const factory Measure.distance(
    Distance field0,
  ) = Measure_Distance_Variant;
}

...
abstract class Measure_Speed_Variant implements Measure {
and various others
```

but there is no Measure_Speed.

So I guess you can directly remove _Variant, and run codegen and everything (maybe previously we forget to run some codegen like freezed?)

If there is error please commit and push the code so I can see it. Currently that two files looks like no problem (if we remove variant suffix)

Reply:
https://github.com/fzyzcjy/flutter_rust_bridge/blob/e63577dd12/frb_example/pure_dart/rust/src/bridge_generated.rs#L1234-L1257

I see. So it is clear now:

```rust
#[repr(C)]
#[derive(Clone)]
pub struct wire_Measure {
    tag: i32,
    kind: *mut MeasureKind,
}

#[repr(C)]
pub union MeasureKind {
    Speed: *mut Measure_Speed,
    Distance: *mut Measure_Distance,
}

#[repr(C)]
#[derive(Clone)]
pub struct Measure_Speed {
    field0: *mut wire_Speed,
}

#[repr(C)]
#[derive(Clone)]
pub struct Measure_Distance {
    field0: *mut wire_Distance,
}
```

We need to change the Measure_Speed into wire_Measure_Speed.
Notice that wire_Measure is correctly named (it has wire in it), but not Measure_Speed (should be wire_Measure_Speed)

(The discussion spanning too long time makes me forget everything from the memory... So sorry for confusion)

Reply:
> (The discussion spanning too long time makes me forget everything from the memory... So sorry for confusion)

No worries, same here 😅 

Ok I tried adding the `wire_` prefix in the latest commit and pushed the codegen altogether @fzyzcjy

Reply:
Seems almost finished now :)

```
error[E0412]: cannot find type `Distance_Unknown` in this scope
  --> src/bridge_generated.rs:46:19
   |
46 |     Unknown: *mut Distance_Unknown,
   |                   ^^^^^^^^^^^^^^^^ help: a struct with a similar name exists: `wire_Distance_Unknown`
...
52 | pub struct wire_Distance_Unknown {}
   | -------------------------------- similarly named struct `wire_Distance_Unknown` defined here
```

Just adding `wire_` to them and I guess we can merge it

Reply:
done in these commits [6d51df0](https://github.com/fzyzcjy/flutter_rust_bridge/pull/604/commits/6d51df092b1f364f7e9f8950eb2e830794293ded) and [ebdfd81](https://github.com/fzyzcjy/flutter_rust_bridge/pull/604/commits/ebdfd81613b7f2dbcaf5b029a698bd372cd74908)

Reply:
still outputting errors everywhere though 😓

Reply:
almost 4k lines of code disappears... this may be the problem

![image](https://user-images.githubusercontent.com/5236035/184080778-d41da1a3-28b5-4491-9720-b8a9e6460e62.png)

Reply:
Yes, any idea why?

Reply:
oops your api.rs also lose a lot

![image](https://user-images.githubusercontent.com/5236035/184087473-0d256f76-e4ee-4337-a57d-b16baea03ebe.png)

Reply:
Hmm so what's next ?

Reply:
ensure api.rs is right :)

Reply:
I don't mind digging further but tbh I won't have much time to allocate in the coming weeks so I'm worried it might go stale. If you want to take the lead and finish it I wouldn't mind!

Reply:
Well, then maybe we should just leave it the original name and I will merge it.

Reply:
utterly agree! do you want me to force push to revert the latest commits?

Reply:
It depends on you :) Git also has a feature called "revert" as well

Reply:
I might have found the missing bit, let's see CI outcome

Reply:
@fzyzcjy looks like we can have the cake and eat it too 🍰 😄